### ISSUE TITLE
Add an option to fix the Z-forward vector when importing a GLTF file 

### ISSUE THREAD

**Issue by geowarin** [godotengine/godot-proposals#6198]
### Describe the project you are working on

A 3D Game.

### Describe the problem or limitation you are having in your project

The GLTF specification says:

> glTF uses a right-handed coordinate system, that is, the cross product of +X and +Y yields +Z. glTF defines +Y as up. The front of a glTF asset faces +Z.

Which, for some reasons, contradicts Godot, which recommends that -Z be the forward direction.

In blender, if I have a character set up the conventional way:

![Front view with -Y forward](https://user-images.githubusercontent.com/1045636/215348830-c5b50064-15b8-4e6a-8c3f-42f4be7265dd.png)
*Front view with -Y forward*

I export it as gltf and import it into godot, my character is facing the wrong way.

![Front camera is showing the back of my character](https://user-images.githubusercontent.com/1045636/215348871-e0d0efb0-c638-4722-ba15-b489766ab3e3.png).
*Front camera is showing the back of my character (he is looking towards +Z)*

![image](https://user-images.githubusercontent.com/1045636/215479923-1048c276-9f17-4f9d-a78b-120c51bf7f2b.png)

This is what I found about GLTF decision:

- https://github.com/KhronosGroup/glTF-Blender-IO/issues/712
- https://github.com/KhronosGroup/glTF/issues/1043

--- 

This is not a discussion about which convention is the "right one". 

Changing the Z direction (or any other axis) in the editor has already been discussed [multiple](https://github.com/godotengine/godot-proposals/issues/545) [times](https://github.com/godotengine/godot/issues/17872#issuecomment-377691846) and rejected. This is a core design decision of Godot.

---

I don't want to manually flip all my characters so that they are facing the right way in Godot.
The vast majority of 3D tools with GLTF export will consider +Z as the forward direction.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I suggest adding an import option for GLTF, FBX  and blend files (they are all essentially importing GLTF) that allows users to flip the meshes and skeletons by 180° around the Y axis.

It should be convenient for users to use this as the default for all 3D imports.

Maybe this should even be the default behavior, otherwise meshes will be facing backwards most of the time.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

A checkbox in the importer labelled "-Z is forward (godot convention)" or something like that.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

One could write a custom import script that flips the root node.
But I think it is a question of having good defaults: using any model generated by 3D tools that export to GLTF will face backwards in the editor.

### Is there a reason why this should be core and not an add-on in the asset library?

This reduces confusion when people use default methods like `look_at` and copy code from official tutorials, while using the vast majority of available 3D assets.

**Comment by geowarin** [comment]
At the very least, [the documentation](https://docs.godotengine.org/en/latest/tutorials/assets_pipeline/importing_scenes.html#exporting-gltf-2-0-files-from-blender) about importing from blender should acknowledge the problem and five user a viable solution.

**Comment by TheDuriel** [comment]
Uhm...

In Blender, your character is looking towards -Y.

In Godot, your character is looking towards -Z.

This is consistent across both and the GLTF spec you quoted. We only swapped letters, not directions.

The difference here is not that the character is rotated in an unexpected direction, but that the camera is on the other side. The editor camera is placed at a positive Z coordinate, while in Blender it is placed at a negative Y coordinate. There is nothing _functionally_ wrong here.

And to me this makes sense, since for the purposes of gameplay we want to consider viewing a character from their back side. Not just player controlled models.

**Comment by TheDuriel** [comment]
Maybe the real "issue" here, is that the camera defaults to the wrong location.

**Comment by geowarin** [comment]
Nope. Character is looking towards +Z in Godot. That's the problem.

![image](https://user-images.githubusercontent.com/1045636/215479923-1048c276-9f17-4f9d-a78b-120c51bf7f2b.png)

**Comment by TheDuriel** [comment]
Huh, can confirm.

I was under the expression from my past uses of Godot 3D that this was not the case.

In that case though, no need for a proposal. That's a straight up bug. The importer is actively rotating the model, already.

**Comment by geowarin** [comment]
@TheDuriel Are you sure this is a bug? I've looked quickly at the importer code and I see no extra rotations being applied other than the one described in the GLTF file.

This seems to be the core of the issue: conflicting conventions between gtlf and Godot

**Comment by Flavelius** [comment]
I have a scene that relies on planar top down worldspace projection. This is a blender file, but i think it's related, as the importer afaik converts it internally to gltf. As of now the model imports exactly in the same position and rotation as it is in blender, which is fine in my case. This currently only requires me to add an offset in blender (to -z space) and to the projection logic there. If there was automatic axis flipping I'd have to add that to the Godot shader and logic aswell, which would not be ideal.
In my opinion axis conversion should be (added as) an import option, as both ways are valid.

**Comment by fire** [comment]
Historically as the group who proposed adding this to Blender we think the fortune algorithm and the temperance algorithms are a bad idea because it breaks bones. https://github.com/KhronosGroup/glTF-Blender-IO/issues/1662 @lyuma Also I think we’ve implemented this 3-6 tines jn Godot Engine.

**Comment by jtnicholl** [comment]
I think it'd be better if Godot followed convention and treated +Z as forward.
Fixing the editor camera would be simple, though changing methods like `look_at` would break compatibility.

**Comment by TheDuriel** [comment]
I do not understand Z+ as forwards to be a common convention at all.

**Comment by Calinou** [comment]
> I think it'd be better if Godot followed convention and treated +Z as forward. Fixing the editor camera would be simple, though changing methods like `look_at` would break compatibility.

There is no set convention for 3D software: https://docs.godotengine.org/en/latest/tutorials/3d/introduction_to_3d.html#coordinate-system

**Comment by Flavelius** [comment]
> I do not understand Z+ as forwards to be a common convention at all.

It's easier for humans to visualize because when you look top down x & y(then z as up-direction) will still be positive and projections and math you do in this view don't have to be value-flipped on one axis to map to memory indices. It also practically aligns with how humans also tend to associate negative with backwards and positive with forwards, as in time and speed which maps perfectly to motion and thus vectors.

**Comment by TheDuriel** [comment]
A camera inherently views _into_ a space. That space is negative, since it is _inside_ your monitor.

There is an equivalent amount of analogies that can be made to explain either direction as the default.

From a technical standpoint, Z- is correct for cameras. And since cameras look "forwards", that axis should translate to your character models.

**In any case** This issue is about the fact that: Y- becomes Z+ in Godot, when it should be Z-.

Simple as that.

**Comment by jtnicholl** [comment]
> There is no set convention for 3D software: https://docs.godotengine.org/en/latest/tutorials/3d/introduction_to_3d.html#coordinate-system

Maybe "convention" is not the best term, but it's the most common. (At least, right-handed Y-up is, not as sure about +Z being forward.) And the fact that it's not completely standardized is very annoying, I get tripped up often going between Blender and Godot.

Besides, Godot is not even consistent about Z+ being forward, for example the advanced scene import window opens showing the Z+ side of the scene facing the camera. If Z- is forward that means it's opening backwards.

> **In any case** This issue is about the fact that: Y- in GLTF becomes Z+ in Godot, when it should be Z-.
> 
> Simple as that.

No? The GLTF specification states that Z+ is forward, not Y-. This is directly quoted in the proposal. If you mean _Blender_ is converting Y- to Z+, that's what it's supposed to do.

The fix is to either rotate imported GLTF files 180 degrees around the Y axis, or treat Z+ as "forward". I was just saying that I would prefer the latter.

**Comment by TheDuriel** [comment]
The fix is to indeed, to offer an option to rotate the axis upon import, and to enable it by default. So that the model does not flip from blender to godot.

No change to godots forward axis needs to be made at any point. It's completely irrelevant that it is Z-. Literally, you will never interact with that fact while you are modelling, animating, or programming your game. In the former cases, the perceived difference should be handled by the export/import steps. And in the latter case you are working with the FORWARDS constant provided by Godots various types.

Lets not make this proposal about something it's not.

An alternative Fix would be to advise Blender users to configure their axies before export, something blender is fully capable off. And which is a standard step in asset production targeting specific formats. But Godot should be _nice_ enough not to ask of you.

**Comment by jtnicholl** [comment]
> It's completely irrelevant that it is Z-.

If it didn't matter, we wouldn't be discussing it.

> Lets not make this proposal about something it's not.

The purpose of proposals is to make a suggestion, and have a discussion on whether it should be implemented and how. Mentioning an alternative is not off-topic.

> An alternative Fix would be to advise Blender users to configure their axies before export, something blender is fully capable off. And which is a standard step in asset production targeting specific formats. But Godot should be _nice_ enough not to ask of you.

I just said that Blender's GLTF exporter behaves as it should. Nothing about that part of the pipeline should change, if it did it wouldn't be complying with the GLTF standard.

**Comment by TheDuriel** [comment]
It doesn't matter that it's Z- or Y+. It matters that the model is facing BACKWARD.

**Comment by QbieShay** [comment]
It seems to me that there are enough points of agreement here to converge this conversation:
1. The models are currently facing backwards. This is an issue.
2. Godot uses -Z forward and we do not want to change that (it's a really gigantic change to do and it 100% not happening)
3. The likely simplest fix for this would be to offer an optional checkbox upon import, off by default, which would keep consistency.

If there's any other point, feel free to add, but otherwise I think that discussing adding Z+ as a forward to Godot is completely out of scope.

Also, it seems like some of you got a bit heated during this discussion. I encourage you to take a pause if that's the case.

**Comment by QbieShay** [comment]
Also please note that this behaviour has been the case for very long. Comments including "Godot should just do X" downplay the complexities involved when dealing with different space coordinates and skinned meshes. Please be kind and patient.

**Comment by geowarin** [comment]
Changing the Z direction (or any other axis) in the editor has already been discussed multiple times and rejected. This is a core design decision of Godot.

For instance, see this and the multiple issues it relates to:
- https://github.com/godotengine/godot-proposals/issues/545

To [quote reduz](https://github.com/godotengine/godot/issues/17872#issuecomment-377691846):

> This should be closed and any further requests to change how axis work should point to this issue

---

> Comments including "Godot should just do X" downplay the complexities involved when dealing with different space coordinates and skinned meshes. Please be kind and patient.

This is interesting, I am not real familiar with the importer, can you explain what kind of problems you anticipate for this feature to be implemented?

**Comment by YuriSizov** [comment]
> This is interesting, I am not really familiar with the importer, can you explain what kind of problems you anticipate for this feature to be implemented?

I don't think it was in reference to the importer. Adding a flag to the importer should be pretty straightforward (though don't quote me on that, file formats for 3D models are a special kind of hell). Qbie's request was more general to the discussion about what _should_ be the correct way, as if this is not a cornerstone of entire rendering, animation, importing, etc systems. Saying that this _should_ be changed to another coordinate system is downplaying the profound effect of such a change on the entire engine.

**Comment by geowarin** [comment]
Yes sorry, I misread.

So real question: if we decide to implement this feature, should it be the default?
I *will* break all the games that manually flipped imported GLTF or worked around `look_at` facing backwards.

So if we want to make this the default, it should probably happen before 4.0 is released?

**Comment by jtnicholl** [comment]
The already rejected proposals that were linked are suggesting a change away from a right-handed Y-up coordinate system, which is not what I was suggesting. I was only suggesting to change the _names_ of the directions. That is a much smaller change.

If even that is too big especially in time for 4.0 that's understandable, and it is perhaps getting a bit off-topic, but I wanted to make it clear what I said.

**Comment by oenixphay42** [comment]
What about replacing the Root Scale importer setting with a Root Transform? It would solve all of the issues like this without adding settings for compatibility with specific formats.

**Comment by Flavelius** [comment]
> What about replacing the Root Scale importer setting with a Root Transform? It would solve all of the issues like this without adding settings for compatibility with specific formats.

With Transforms you only have rotation and scale available for adjustments where scale introduces other issues (with colliders, mesh normals and probably more), and using rotation results in the x axis being flipped too, so this likely won't be enough.

**Comment by jtnicholl** [comment]
> With Transforms you only have rotation and scale available for adjustments where scale introduces other issues (with colliders, mesh normals and probably more), and using rotation results in the x axis being flipped too, so this likely won't be enough.

Just rotating would be enough for GLTF, it doesn't just treat Z+ as forward, but also X+ as left. 
Although, there's other formats which don't have their coordinate systems 100% standardized, so having transform options may be useful for those. We can flip the normals again after scaling, and apply colliders afterwards, so I don't think that's an issue. I can't think of anything else that would be a problem.

**Comment by Flavelius** [comment]
> Just rotating would be enough for GLTF, it doesn't just treat Z+ as forward, but also X+ as left.

Interesting, i didn't know and expect that it defined +x as left; Coordinate space conventions truly are a wild west.
 
> I can't think of anything else that would be a problem.

I also remember animations as something that breaks when negative scale is _applied_ (even sometimes does in unity these days, when exported from blender)

**Comment by TokageItLab** [comment]
It is a glTF specification that the asset must be oriented in the +Z direction.

https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#coordinate-system-and-units

So I assure that importing the -Z axis as forward in Godot's importer will never be supported in the core.

If you absolutely need it, you just write it as an import plugin.

**Comment by TheDuriel** [comment]
A forward facing model in GLTF, should arrive forward facing in Godot. Without an additional rotation being applied by the user.

To be clear, models are not arriving flipped. They are arriving facing 180° from where they are expected to be.

The cleanest way to ensure that models arrive facing the same direction they were exported in, is to allow the user to specify the forward direction in the importer. Since the importer will need to rotate the entire stored scene around the origin.

A post import script via editor plugin can't handle that.

**Comment by TokageItLab** [comment]
PathFollow3D also treats +Z as forward. Actually, what are you saying about Godot treating -Z as forward? If it exists, we should fix that method.

**Comment by TheDuriel** [comment]
https://docs.godotengine.org/en/stable/classes/class_vector3.html#:~:text=Down%20unit%20vector.-,FORWARD,-%3D%20Vector3(%200

Vectors, Transforms, Cameras, and Spatial Rotations, all treat Z- as forwards.

This would imply that PathFollow3D is inverted.

**Comment by jtnicholl** [comment]
The editor camera should default to looking backwards, that means it's looking at the front of the scene.

**Comment by TokageItLab** [comment]
Indeed... Personally, I think Vector3.FORWARD should be +Z, but I don't know how much of an impact that would cause. But, as long as we keep the right hand coordinate system, I don't think that many modifications will be necessary.

However, most importantly, we are now just before 4.0 RC. To change it now is quite horrible.

As for the camera facing backward by default, I don't see a problem for viewing the asset from the front. If the asset is to be imported with -Z as forward, then the default camera placement would need to be reversed as well.

**Comment by TheDuriel** [comment]
SO ANYWAYS.

Now that all the newcomers to the thread are up to speed on Godots forward direction, and the **unchangeable nature of it.** (Scroll up.)

The GLTF Importer is applying a 180° rotation to imported models. This is wrong. It needs changing. And doing so **before** this gets locked in for the entire lifetime of 4.x is probably a good idea.

Because if not now. Then when? Certainly we won't be breaking everyone's 4.0 projects with 4.1.

**Comment by geowarin** [comment]
A little off topic but PathFollow3D is indeed inverted: https://github.com/godotengine/godot/issues/53837

**Comment by jtnicholl** [comment]
People misinterpreted/misread what I wrote. What can't be changed is the right-handed Y-up system. Switching forwards with backwards and left with right hasn't been proposed before (AFAIK). While I still agree it's probably too big to do right before the RC, it doesn't require changes to the underlying math, just some renaming. And Godot is already inconsistent with these labels anyways.

Though for now, we should do something that doesn't break compatibility. Making Z+ forward can be discussed in a separate proposal for 5.0.

Also, please stop shouting at people in all-caps and bold text.

**Comment by TokageItLab** [comment]
> it doesn't require changes to the underlying math, just some renaming.

I agree.

I have worked primarily for animating glTF assets. But to be honest, I never noticed that -Z is labeled Forward until now. Since the asset needs to go forward, I had no idea that the PathFollow3D direction was the other way around.

Yes, I think it is possible to just rename it.

Whether good or bad, core uses vector with direct numerical values inside core instead of using constants such as Forward, so the renaming will not affect core's behavior. (However, it may destroy the user's project)

If do that, I feel it is strange that the camera and lights are backward facing, but that can be fixed in 4.1 or 5.0 in the future.

**Comment by geowarin** [comment]
Ok so in this case, we should make the front view to be Z+, [which was changed on Feb 2021](https://github.com/godotengine/godot/pull/45669) and fix `look_at`, which considers Z- to be forward.

Funnily enough, on the topic of look_at, people were having a[ very similar conversation](https://github.com/godotengine/godot/issues/27544) to the one we are having right now.

**Comment by TheDuriel** [comment]
I don't know why we are having this conversation whatsoever. It does not relate to the problem at hand.

**Comment by geowarin** [comment]
> I don't know why we are having this conversation whatsoever. It does not relate to the problem at hand.

Well, we assumed that changing the forward direction would be a massive undertaking but if it is only:

- Changing some constants
- Changing the front view
- Changing the `look_at` function

Then we can have the same convention as GLTF and thus we are solving issue in a much more elegant way.

**Comment by TheDuriel** [comment]
> Then we can have the same convention as GLTF

Which was never the question.

And does mean that when, after this change, Godot does not follow the forwards standard for FBX, DAE, USD, or other potential formats, we then go and change Godots forwards direction again?

When in fact. The _problem at hand_ is the lack of an applied rotation in one specific asset import script?

**Comment by TokageItLab** [comment]
If the forward vector can be set by the user in the import or project settings, the forward vectors of some Nodes and method must also be changed at the same time, as mentioned above.

So, systemically, "only one vector is forward" is straightforward.

Surely, the problem is that Godot currently defines -Z as forward, but loads the forward of the glTF asset as +Z. This can be solved by loading the forward of the glTF asset as -Z.

But in this case, loading the forward of the glTF asset as +Z should not be an option, for the reasons stated above.

---

However, in the opinion of myself, @jtnicholl and my modeler friends, it is minority that -Z is forward in the right handed Y-UP coordinate system. Walking animations are usually created so that it approaches the default camera.

The default camera shots the asset's face. And it is common to use the side with the asset's face as the Forward. However unfortunately, Godot uses the asset's hips as the forward vector.

In summary, the confusion is caused by the fact that the Godot's default camera direction, glTF importer and PathFollower3D use +Z as the Forward, but Godot's Vector3, Camera3D, SpotLight3D Forward is labeled as -Z.

---

> And does that mean that when, after this change, Godot does not follow the forwards standard for FBX, DAE, USD, or other potential formats, The we then go and change Godots forwards direction again?

It might happen if a right-hand coordinate format emerges that Godot supports in preference to glTF, but I guess that the chance is almost zero.

I don't know of any format that defines forward as clearly than glTF. Also, since Godot is the most supported glTF, so I suggest that +Z should be forward to be consistent with the glTF specification.

**Comment by TheDuriel** [comment]
> the problem is 

That assets imported by the GLTF scene importer, face the wrong direction.

**Comment by TokageItLab** [comment]
> That assets imported by the GLTF scene importer, face the wrong direction.

That cannot be said without presuming that Godot assumes -Z to be Forward. In fact, Godot is trying to make +Z a Forward in some places, as noted above, so it's not just about the glTF importer.

**Comment by TheDuriel** [comment]
Godot has made it unquestionably clear as to which direction it considers forwards. The fact that, a node or two, get it wrong, does not invalidate this. It instead means that _someone needs to write an issue to correct this mistake._

Assets imported by the GLTF importer, face Vector3.BACKWARDS as per the Vector3, transform, basis, and other constants. Necessitating user intervention if the user wishes to move the asset Vector3.FORWARDS.

Other importers make an assumption about what axial system the format uses, and by extension of the original problem that this thread is about, may require additional import settings to be added. That is not in the scope of this proposal however. This proposal is about the fact that the GLTF importer needs to rotate assets 180°. Which is not controversial. And is not a breaking change.

**Comment by Flavelius** [comment]
> > That assets imported by the GLTF scene importer, face the wrong direction.
> 
> That cannot be said without presuming that Godot assumes -Z to be Forward. In fact, Godot is trying to make +Z a Forward in some places (default camera direction, glTF importer and PathFollower3D), as noted above, so it's not just about the glTF importer.

Even the documentation states that +z is 'Front' and forwards and only that cameras 'Look 'Back':
https://docs.godotengine.org/en/stable/tutorials/3d/using_transforms.html#introducing-transforms
which even coincides with the basis identity

**Comment by TheDuriel** [comment]
Tutorials should be taken with a grain of salt. This one is simply wrong. If you orient your characters to face -Z, then they will walk backwards if you use any of the built in methods without an additional rotation.

Taking the quote "the camera looks back" as being indicative of the entire engine treating Z- as backwards is not a good argument either.

**Especially** when the tutorial later points out the correct direction. https://docs.godotengine.org/en/stable/tutorials/3d/using_transforms.html#introducing-transforms:~:text=Imagine%20you%20need%20to%20shoot%20a%20bullet%20in%20the%20direction%20your%20player%20is%20facing.%20Just%20use%20the%20forward%20axis%20(commonly%20Z%20or%20%2DZ).

**Comment by Flavelius** [comment]
Then again if you take the z component from Basis.Identity as in the shoot example, the bullet will fly towards +z, which is expected as the identity basis should have forward semantics. Only if used from the camera which looks 'backwards' does it fly in the camera's forward

**Comment by aaronfranke** [comment]
The standard rotation for Y-is-up right-handed coordinate systems is that +X is right and +Z is back (-Z is forward).  
The standard rotation for Z-is-up right-handed coordinate systems is that +X is right and +Y is forward (-Y is back).  
-Z being forward is implied if we accept these two truths first:

* +Y is up
* +X is right

If we spun 180 degrees around Y to keep +Y up to make +Z be forward, then that would also imply that +X becomes left and -X becomes right. That's silly, it is directly contrary to 2D and most expectations (and yeah Godot's 2D has +Y down but uh let's ignore that...). Users expect +X to be right.

---

Now, let's talk about how this relates to Blender -> GLTF -> Godot:

Godot and Blender both use the same handedness, right-handed. The difference is that they are rotated 90 degrees around X relative to each other. This is why +Y in Blender is -Z in Godot, and -Y in Blender is +Z in Godot. This is the correct mapping.

GLTF itself uses Y-is-up right-handed, the same as Godot. The program doing the conversion is Blender. Blender is (correctly) mapping its +Y to GLTF -Z and its -Y to GLTF +Z. @TheDuriel Godot is NOT "applying a 180° rotation to imported models". Neither is Blender.

If we want to add an option to the GLTF importer to spin the model 180 degrees, that's fine, but that's an extra feature we are adding. It's not a bug. The models are being imported exactly as-is. Godot is not doing any conversions whatsoever right now. We can just add a bool/checkbox to the importer that does this (anyone want to bikeshed on the name? something like "rotate 180 degrees around Y"), which will default to false, just like the OP proposes.

> Even the documentation states that +z is 'Front' and forwards and only that cameras 'Look Back':
https://docs.godotengine.org/en/stable/tutorials/3d/using_transforms.html#introducing-transforms

Sorry that this has caused you confusion. This article is really bad, it needs to be rewritten from scratch. I've known about it for years, but I keep forgetting about it...

**Comment by Flavelius** [comment]
I did a few more tests in godot and blender and i stand corrected that the forward semantics atleast are consistent between both, (which then applies to gltf aswell), through models and shader logic:

![grafik](https://user-images.githubusercontent.com/8841352/216816616-d48d8df6-5fc5-4e1f-b5dc-4abdefc88db4.png)

[BlendImportAxis.zip](https://github.com/godotengine/godot-proposals/files/10610897/BlendImportAxis.zip)

This makes it a little less of an issue for my case. It's still some (mental/performance) overhead for flipping z in code for the rest of the worldspace logic, though not ideal, i can live with that; it's the better alternative for me, than flipping things in blender and in shader

**Comment by QbieShay** [comment]
> This is interesting, I am not really familiar with the importer, can you explain what kind of problems you anticipate for this feature to be implemented?

@geowarin 3d coordinate spaces are a bit messy. Say for example, in the importer we just flip every z coordinate. Cool right? it should be just what we want and what we've been discussing ..?

Not quite. To understand this you have to understand convention for cross products and coordinate spaces.  @aaronfranke gave an in-depth explanation, but I'll illustrate the bases (pun intended)

Freya holmer has a nice diagram about different coordinate systems: https://twitter.com/FreyaHolmer/status/1325556229410861056?lang=en

You could deduce a 3D basis ( imagine it as a point of reference of a 3D space ) just by two vectors, instead of three, because the third vector can be inferred by making a cross product of of the first two, **in the right order**. Which way the third vector points, depends on the convention that you choose. Left hand or right hand (you put first vector on the thumb, second on the index, and then you point your middle finger away from your palm: you'll see that if you contort a bit and align both your thumbs and your indexes, your middle fingers will point in opposite direction for your two hands).

Flipping the Z axis, as-is, means changing the entire coordinate system convention for that model. Because Godot uses another coordinate systems, then interpreting rotations of the bones and the way they skin to the mesh will become really problematic. You'd end up with bones rotating in funny ways around axis you don't expect and your character being most definitely in pain :D

Like aaronfranke pointed out, simply rotating the model does still create some inconsistencies on what's right and left in the model. Likely, it can be still worked with, but i hope this comment sheds at least a bit of light of why it's not something that can be fixed just by snapping fingers.

> will it be default

I don't think it should be on by default, otherwise everyone that has imported models now will have them flipped around. It can however be a project setting to be set as a default, so it needs to be changed only once per project, which to me seems ilke a reasonable compromise.

**Comment by TheDuriel** [comment]
Godot 4 is not out yet. If this were to make it in, it could be one by default. And disabled by the project converter.

Realistically it's not going to make it in.

> Flipping the Z axis, as-is, 

Would result in the model ending up mirrored as well. So it needs to be rotated.

**Comment by QbieShay** [comment]
> Godot 4 is not out yet. If this were to make it in, it could be one by default. And disabled by the 3.x to 4.x project converter.

Godot4 is in beta stage, no more compat breaking changes.

**Comment by aaronfranke** [comment]
I opened a PR to add an import option for this, took under 2 hours (needs testing): https://github.com/godotengine/godot/pull/72753

Also worth noting is how GLTF is not self-consistent with how it defines forward. [The camera spec](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#view-matrix) defines forward as -Z (as a behavior of [the camera's orientation](https://github.com/KhronosGroup/glTF-Tutorials/blob/master/gltfTutorial/gltfTutorial_015_SimpleCameras.md)), while [the coordinate spec](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#coordinate-system-and-units) defines +Z as forward (as a convention).

<details>

<img width="800" alt="Screenshot 2023-02-05 at 5 48 35 AM" src="https://user-images.githubusercontent.com/1646875/216816941-2bbeaf7c-8cc9-4e66-ad8d-a91375ad57b1.png">

</details>

At a glance, this may seem fine, and it works for placing a camera in silly little scenes with a single prop. However, when you think about it more, if you want two objects to face each other - doesn't that mean that one of them *should* have a transform rotated the other way? Going further, it completely breaks down for the cases you actually use cameras for, where you expect cameras to face the same direction as the assets, like for player characters (I rotated mannequiny):

<details>

<img width="600" alt="Screenshot 2023-02-05 at 5 56 09 AM" src="https://user-images.githubusercontent.com/1646875/216817272-bfaddd45-81a8-4c2f-8b67-baeeb09a54b5.png">

</details>

Also worth noting is how GLTF is not self-consistent with how it defines scale. GLTF defines the unit of distance as meters, but boom box sample asset used in the spec is 0.02 units in diameter, which would make it 2cm. That's way too tiny for a boom box. The inconsistent conventions leads me to believe that the person who wrote section 3.4 of the GLTF spec, and made the boom box, was not communicating correctly with the writers of the other parts of the spec. Or maybe it was ruined by a committee settling on an inconsistent compromise.

Out of curiosity, I also checked how Blender defines cameras. They... point down by default. Interesting choice.

Godot is self-consistent, the rest of the world is not. So now we have to add compatibility options. ¯\\\_(ツ)\_/¯

**Comment by QbieShay** [comment]
> Godot is self-consistent, the rest of the world is not. So now we have to add compatibility options. ¯\_(ツ)_/¯

I think what's important for Godot is to add ease of use, and rotating all your models 180degrees and adding a pivot is something that (imo) Godot could take care of, since it's something that's consistently a hassle across 3D projects.

**Comment by QbieShay** [comment]
After a lengthy discussion with other maintainers, it is clear that this issue is not as simple as it appear. Thanks to @aaronfranke 's attempt at a rotation, i think it's obvious why just rotating the object doesn't work. Importer maintainers have voiced their concern over adding complexity to the importer by adding extra rotations, and other maintainers have voiced concern over possibly adding extra convetions or look_at methods and so on.

It's clear to us that this is an involved discussion that needs to happen in a moment that does not have a massive release coming which is already in feature freeze.

To the people experiencing this as an issue: please be patient. You can write your own post-import script that adds an extra node on top of the hierarchy to rotate the model. Once we settle on a solution for this, we'll update this thread.

For now, there's no point in continuing the discussion. We need to take our time and gather our ideas on how to best address this in the best way possible and with the lowest impact possible on the various areas.

To Godot users pitching in this thread, thanks for your feedback. It has sparked a lot of discussion and allowed us to look deeper into things that turned out to be more complex than we thought.

To maintainers participating, thanks for putting your energy into this. Your input has been very valuable and allows people reading to understand the complexity of the topic at hand.

**Comment by fracteed** [comment]
Sigh, I really wish this was never changed in the first place, since it worked fine in early 3.x versions. People can argue semantics about the which way front is based on the axis, but having a front view should adhere to the way it works in every other 3d package out there.

I have used most 3d packages over the years and regardless of their worldspace up axis, they all have the front view based on what you see in the top view. If a character in the top view is looking down (in screen space), then this is the direction that the front camera should be looking towards. As can be seen from this screenshot in blender. I have put the arrow in to indicate this direction. This is what artists expect and is the way that most people would model and rig a character. It doesn't matter if it is a y-up or z-up package or negative/positive in the view direction.

![front_down](https://user-images.githubusercontent.com/7741669/216929498-1308bbf9-29da-4869-b3f8-67d010955ee3.jpg)

I currently have my front hotkey mapped to rear in godot, so I have learned to just ignore what the godot viewport label says. The best way of solving this issue is to make an option in the preferences so that front view can be set to either positive or negative z direction. Surely, this would keep everyone happy?

Having an option to rotate 180 degrees on import is just going to cause more confusion and is an attempt to correct a bad default. Most Godot users are going to model their assets in Blender, Maya or Zbrush and they all adhere to what I describe above.

**Comment by fracteed** [comment]
@QbieShay not sure as I was using dae back then. What I am referring to is the way that the front ortho camera works in Godot which has changed since then. It used to be consistent with the way Blender or Maya, etc worked.

As far as i can tell, the OP is having issues with what front ortho looks like in Godot, as it is the reverse of the conventional way (as used by blender). The model has been imported (presumably) correctly into Godot, but the Godot front camera is what a rear camera is doing in Blender. This is why I was suggesting having a settings option to orientate the front ortho camera in Godot to be the same as Blender. This way the front camera is consistent between Blender and Godot and is what most artists would expect.

**Comment by aaronfranke** [comment]
A short summary of PRs working on fixing this problem in different ways:

From @aaronfranke:

* Add an import setting to rotate by 180 degrees. A small PR, only 25 lines. Does not break compat. Not mutually exclusive with any of the other PRs. https://github.com/godotengine/godot/pull/72753
* Add a bool to `look_at` and `looking_at` to make it use +Z instead of -Z. Off by default. Does not break compat. This is a subset of @TokageItLab's PR. Conflicts with @reduz's PR. https://github.com/godotengine/godot/pull/75875

From @TokageItLab:

* Add a bool to `look_at` and `looking_at` to make it use +Z instead of -Z, and add `GLTF_*` constants to represent the directions when +Z is forward. Off by default. Does not break compat. This is a superset of @aaronfranke's PR. https://github.com/godotengine/godot/pull/72795
* Flip the meaning of "Front" and "Rear" in the 3D editor. Improves consistency with Blender and FreeCAD but is no longer self-consistent and is inconsistent with Unity and Unreal. https://github.com/godotengine/godot/pull/76052
* Allow changing the 3D editor viewport to have +Z as forward. Allows using either forward direction and is self-consistent in either mode. https://github.com/godotengine/godot/pull/75921

From @reduz:

* https://github.com/godotengine/godot/pull/76060
    * Define two `look_at` methods with two forward directions, one for the camera, and one for models (the model one uses the naming "front"). This also implies two back directions (model is "rear"). Breaks compat but old projects will keep working since deprecated constants and methods are kept.
    * A previous version of the PR also defined two left directions and two right directions. The PR has since been updated to only add the methods and the front/rear constants but no extra left/right constants, so if you use the model front/rear constants then they are not consistent with the existing left/right constants.

**Comment by Zireael07** [comment]
@des1redState A workaround is to parent your model to an empty Spatial and rotate said empty spatial to however you want

**Comment by aaronfranke** [comment]
All PRs addressing this issue have been either merged or closed, and as far as I know all engine developers are satisfied with the current state. Therefore, I am closing this PR as resolved.

In addition to the list of engine PRs I made above, the 3D asset direction conventions have been documented on the [Importing 3D scenes page](https://docs.godotengine.org/en/latest/tutorials/assets_pipeline/importing_scenes.html#d-asset-direction-conventions). This page describes briefly, but dense with detail, how local and global directions work in Godot, Blender, and GLTF.

I have some closing notes to hopefully help clarify things. Much of this is focused on terminology and is redundant with the documentation. Pay attention to the terminology: "forward" vs "front".

* Cameras intrinsically only have local directions, not sides, since they are points that view the world in a direction. Therefore it makes more sense to talk about the "forward" and "back" instead of "front" and "rear". To clarify this, you can use the terminology "camera forward" and "camera back". There is also the terminology of "camera left" vs "camera right", flipped compared to model sides.

    * However note that many other parts of Godot use the same directions as cameras, such as lights and audio players, so "camera" may be a slightly misleading clarification.

* 3D models intrinsically have sides. Therefore it makes more sense to talk about the "front" and "rear" instead of "forward" and "back". To clarify this, you can use the terminology "model front" and "model rear". There is also the terminology of "model left side" and "model right side", flipped compared to camera directions.

    * 3D models having their model front be the opposite direction as the camera forward started as a convention to have models face the camera, but it has since been [codified in glTF as part of the glTF 2.0 specification](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#coordinate-system-and-units). Given this rigid standardization, the forward direction of 3D models is not up for debate, and the consensus from the discussion is that Godot should not alter the 3D models on import, so you get exactly what's in the file.

* None of the discussion about local directions/sides is relevant for things like levels/maps, which would instead have cardinal directions like north/south/east/west. Look at the documentation article for more info.

**Comment by TokageItLab** [comment]
Forward is an azimuth and Front is a side of the surface.

Forward indicates the direction in the whole space and is -Z. In this case, Camera front and Light front are -Z because the direction of the camera and lights conform to space.

On the other hand, in OpenGL's right-hand coordinate system, the model is defined as facing the camera, so Model front is +Z.

**Comment by TokageItLab** [comment]
It is certainly model-independent, and one that can look front side of the model when the model is placed at global origin with initial basis.

This allows the user to notice when the user has imported the model in the wrong orientation. Front View is also a convention, and Blender and others have adopted the name Front View.

**Comment by TokageItLab** [comment]
It should almost never be renamed View, as it would be inconsistent with other DCC software and more confusing. Changing Back to Backward might be a good point, but that is all.

**Comment by aaronfranke** [comment]
> Changing BACK to BACKWARD might be a good point (then, LEFT and RIGHT should also be LEFTWARD and RIGHTWARD), but that is all.

That would also imply UP becomes UPWARD and DOWN becomes DOWNWARD.

I'm in favor of using simpler names when possible. We only use "Forward" because "Fore" isn't a common English word.
